validator/client/beacon-api/index.go

```diff
@@ -0,0 +1,80 @@
+//go:build use_beacon_api
+// +build use_beacon_api
+
+package beacon_api
+
+import (
+	"encoding/json"
+	"net/http"
+	neturl "net/url"
+	"strconv"
+
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/v3/api/gateway/apimiddleware"
+	rpcmiddleware "github.com/prysmaticlabs/prysm/v3/beacon-chain/rpc/apimiddleware"
+	types "github.com/prysmaticlabs/prysm/v3/consensus-types/primitives"
+	ethpb "github.com/prysmaticlabs/prysm/v3/proto/prysm/v1alpha1"
+)
+
+func (c beaconApiValidatorClient) validatorIndex(in *ethpb.ValidatorIndexRequest) (*ethpb.ValidatorIndexResponse, error) {
+	stringPubKey := hexutil.Encode(in.PublicKey)
+
+	stateValidator, err := c.getStateValidator(stringPubKey)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get validator state")
+	}
+
+	if len(stateValidator.Data) == 0 {
+		return nil, errors.Errorf("could not find validator index for public key `%s`", stringPubKey)
+	}
+
+	stringValidatorIndex := stateValidator.Data[0].Index
+
+	index, err := strconv.ParseUint(stringValidatorIndex, 10, 64)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to read validator index")
+	}
+
+	return &ethpb.ValidatorIndexResponse{Index: types.ValidatorIndex(index)}, nil
+}
+
+func (c beaconApiValidatorClient) getStateValidator(pubkey string) (*rpcmiddleware.StateValidatorsResponseJson, error) {
+	params := neturl.Values{"id": []string{pubkey}}
+
+	url := buildURL(
+		c.url,
+		"eth/v1/beacon/states/head/validators",
+		params,
+	)
+
+	resp, err := c.httpClient.Get(url)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to query REST API validators endpoint")
+	}
+	defer func() {
+		if err = resp.Body.Close(); err != nil {
+			return
+		}
+	}()
+
+	if resp.StatusCode != http.StatusOK {
+		errorJson := &apimiddleware.DefaultErrorJson{}
+		if err := json.NewDecoder(resp.Body).Decode(&errorJson); err != nil {
+			return nil, errors.Wrap(err, "failed to decode response body validators error json")
+		}
+
+		return nil, errors.Errorf("error %d: %s", errorJson.Code, errorJson.Message)
+	}
```

Comment:
This is a duplicate of code in `getGenesis()`. Given that we will probably use this for every endpoint, please extract a private function. We can parameterize the endpoint name in the first error string.

Reply:
@nalepae I added a helper `getRestJsonResponse` in https://github.com/prysmaticlabs/prysm/pull/1171, you can either copy the implementation or wait until the PR is merged.

Reply:
And vice versa!

Reply:
@PatriceVignola I just tried to merge your `add-get-attestation-data` branch into this one for testing purpose.
Very minor (easily fixables) conflicts (mainly on `BUILD.bazel`, which is expected). 👌

Reply:
I created a dedicated commit in this PR: https://github.com/prysmaticlabs/prysm/pull/11712/commits/52990d21348c055dfcf86be6cc077bc04113eb85
So it's easier for the review to separate your code from #11714 (which I copied here) from mine.